### ISSUE TITLE
Bulk Data Service - java.lang.OutOfMemoryError: Java heap space

### ISSUE THREAD

**Issue by jimk-osu** [ballerina-platform/ballerina-lang#9806]
**Description:**
Running ballerina locally I am consuming another local endpoint that returns 63MB of data and this throws the follow error instantly:

```
java.lang.OutOfMemoryError: Java heap space
```

Ballerina code:
```ballerina
// A system package containing protocol access constructs
// Package objects referenced with 'http:' in code
import ballerina/http;
import ballerina/jdbc;
import ballerina/sql;
import ballerina/io;
import ballerina/log;
import ballerina/config;
import ballerinax/kubernetes;
import ballerinax/docker;
import ballerina/swagger;

// Move the mssql jdbc driver to BALLERINA_HOME/bre/lib
endpoint jdbc:Client simsDB {
    url: config:getAsString("conn.host"),
    username: config:getAsString("conn.user"),
    password: config:getAsString("conn.pass"),
    poolOptions: { 
        maximumPoolSize: config:getAsInt("conn.maximumPoolSize")
     }
};

// Elasticsearch endpoint
endpoint http:Client elastic {
    url: config:getAsString("es.endpoint"),
    timeoutMillis: 180000
};


documentation {
    SIMS `Building` object consists of:
    F{{recordID}} 
    F{{building_Number}} 
    F{{formal_Name}} 
    F{{name}} 
    F{{building_Abbr}} 
    F{{scheduling_Abbr}} 
    F{{address}} 
    F{{city}} 
    F{{state}} 
    F{{zip}} 
    F{{county}} 
    F{{gross_Area}} 
    F{{floors_Above_Ground}} 
    F{{floor_Below_Ground}} 
    F{{investment_Cost}} 
    F{{replacement_Cost}} 
    F{{owner_Name}} 
    F{{region}} 
    F{{zone}} 
    F{{district}} 
    F{{Construction_Year}} 
    F{{renovation_Year}} 
    F{{acquisition_Year}} 
    F{{frame}} 
    F{{facade}} 
    F{{physical_Code}} 
    F{{functional_Code}} 
    F{{maintenance_Provide}} 
    F{{institutional_Type}} 
    F{{framework_District}} 
    F{{pOM_Custodial}} 
    F{{pOM_Maintenance}} 
    F{{pOM_Utility}} 
    F{{campus}} 
    F{{type_}} 
    F{{status}} 
    F{{propertyID}} 
    F{{drawing_Type}} 
    F{{longitude}} 
    F{{latitude}} 

}
type Building record {
    string recordID;
    string building_Number;
    string formal_Name;
    string name;
    string building_Abbr;
    string scheduling_Abbr;
    string address;
    string city;
    string state;
    string zip;
    string county;
    string gross_Area;
    string floors_Above_Ground;
    string floor_Below_Ground;
    string investment_Cost;
    string replacement_Cost;
    string owner_Name;
    string region;
    string zone;
    string district;
    string Construction_Year;
    string renovation_Year;
    string acquisition_Year;
    string frame;
    string facade;
    string physical_Code;
    string functional_Code;
    string maintenance_Provide;
    string institutional_Type;
    string framework_District;
    string pOM_Custodial;
    string pOM_Maintenance;
    string pOM_Utility;
    string campus;
    string type_;
    string status;
    string propertyID;
    string drawing_Type;
    string longitude;
    string latitude;
    string altitude;
};

type Floor record {
    string recordID;
    string name;
    string sort_order;
    string comment;
    string gross_area;
    string _type;
    string campus;
    string building;
};

type Room record {
    string recordID;
    string building;
    string floor;
    string room_number;
    string campus;
    string department;
    string organization;
    string _function;
    string room_Type;
    string sub_room_type;
    string comment;
    string capacity;
    string area;
    string space_ID;
    string _type;
    string POM_custodial;
    string POM_maintenance;
    string POM_utility;
    string assignment_number;
    string effective_date;
    string earning_operation;
};

type Asset record {
    string capacity;
    string headcount;
    string area;
    string vacant;
    string wet_lab;
    string ADA;
    string primary_use;
    string alternate_use;
    string function_category;
    string function_not_used;
    string discipline;
    string space;
    string asset;
    string vendor;
    string manufacturer;
    string model_number;
    string serial_number;
    string location;
    string license;
    string warranty_expiration;
    string warranty_date;
    string install_date;
    string install_cost;
    string reserve_for_replacement;
    string replacement_date;
    string replacement_cost;
    string disposal_date;
    string life_cycle;
    string entity_handle;
    string occupant;
    string asset_ID;
    string asset_class;
    string asset_category;
    string asset_sub_class;
    string asset_name;
    string comment;
    string fixed_asset;
    string phone;
    string picture;
    string campus;
    string building;
    string floor;
    string schedulable;
    string tag;
    string preview;
    string active;
    string drawing;
    string display44;
    string fld_parentID;
    string propertyID;
    string recordDialogID;
    string recordID;
    string spaceID;
};

documentation {
   A service endpoint represents a listener.
}

@kubernetes:Ingress {
    hostname:"someingress.here",
    name:"bal-sims-svc",
    path:"/sims"
}

@kubernetes:Service {
    name:"bal-sims-svc"
}

@kubernetes:Deployment {
    
    image: "somerepo:5000/jimk/simsbal/sims:v1.0",
    name:"bal-sims-svc",
    replicas: 2,
    // Need to build new ballerinax/kubernetes or switch to nightly
    // imagePullSecrets: ["regsecret"],
    enableLiveness: true,
    copyFiles:[
        {
            target:"/ballerina/runtime/bre/lib",
            source:"./conf/mssql-jdbc-6.4.0.jre8.jar"
        }
    ]
}
@swagger:ClientEndpoint 
endpoint http:Listener listener {
    port:9090
};


documentation {
   A service is a network-accessible API
   Advertised on '/sims', port comes from listener endpoint
}

@kubernetes:ConfigMap {
    ballerinaConf:"./conf/sims.toml"
}
// @kubernetes:Secret{
//     secrets: [
//         {
//             name: "bal-sims-db-pass",
//             mountPath: ""
//         }
//     ]
// }
@swagger:ClientConfig {
    generate: true
}
service<http:Service> sims bind listener {
    @swagger:ResourceInfo {
        tags: ["sims", "buildings", "all", "bulk"],
        summary: "Get all SIMS buildings",
        description: "The buildings resource is a bulk call that retireves all buildings in the SIMS system. This should be used sparingly."
    }
    @http:ResourceConfig {
        methods: ["GET"],
        path: "/buildings/",
        produces: ["application/json"]
    }
    documentation {
        ### Get all buildings
        A resource is an invokable API method
        Accessible at '/sims/buildings'
    }
    getAllBuildings (endpoint caller, http:Request request) {
        var selectRet = simsDB->select("SELECT * FROM SIMS.dbo.tbl_SIMS_Building", Building);
        table<Building> dt;

        match selectRet {
        table tableReturned => dt = tableReturned;
        error e => log:printError("Select data from tbl_SIMS_Building failed: "
                               + e.message);
        }

        http:Response response = new;

        var jsonConversionRet = <json>dt;
        match jsonConversionRet {
            json jsonRes => {
                log:printTrace(io:sprintf("%s", jsonRes));
                
                // Untainting return as this resource takes no user input
                response.setJsonPayload(untaint jsonRes, contentType = "application/json");
                _ = caller -> respond(response);

            }
            error e => {
                json errResponse = {
                    "status" : "Connection Error",
                    "error": "500 - Internal Server Error"
                };
                response.statusCode = 500;
                response.setJsonPayload(untaint errResponse, contentType = "application/json");
                _ = caller -> respond(response);
                log:printError(e.message);
            }
        }
     }
    
    // @swagger:ResourceInfo {
    //     tags: ["sims", "buildings", "building", "singular"],
    //     summary: "Get all SIMS buildings",
    //     description: "The buildings resource is a bulk call that retireves all buildings in the SIMS system. This should be used sparingly."
    // }
   
    @http:ResourceConfig {
        methods: ["GET"],
        path: "/buildings/{buildID}",
        produces: ["application/json"]
    }
    
    documentation {
        ### Get specific building by building ID
        A resource is an invokable API method
        Accessible at '/sims/building/buildID'
    }
    getBuilding (endpoint caller, http:Request request, string buildID) {
        
        http:Response response = new;

        // Validate input is an integer
        if(isInteger(buildID)){
            sql:Parameter _buildID = {sqlType:sql:TYPE_VARCHAR, value:sanitizeInt(buildID)};
            
            var selectRet = simsDB->select("SELECT * FROM SIMS.dbo.tbl_SIMS_Building where building_Number=?", Building, _buildID);
            table<Building> dt;

            match selectRet {
            table tableReturned => dt = tableReturned;
            error e => log:printError("Select data from tbl_SIMS_Building failed: "
                                + e.message);
            }

            

            var jsonConversionRet = <json>dt;
            match jsonConversionRet {
                json jsonRes => {
                    log:printTrace(io:sprintf("%s", jsonRes));
                    
                    // Untainting return as this resource takes no user input
                    response.setJsonPayload(untaint jsonRes, contentType = "application/json");
                    _ = caller -> respond(response);

                }
                error e => {
                    json errResponse = {
                        "status" : "Connection Error",
                        "error": "500 - Internal Server Error"
                    };
                    response.statusCode = 500;
                    response.setJsonPayload(untaint errResponse, contentType = "application/json");
                    _ = caller -> respond(response);
                    log:printError(e.message);
                }
            }
        }
        else {
            json json400 = {
                "status": "Invalid building ID",
                "error": "400 - Bad Request"
            };
            response.statusCode = 400;
            response.setJsonPayload(json400, contentType = "application/json");
            _ = caller -> respond(response);
            log:printInfo(json400.toString());
        }
        
    }
    @http:ResourceConfig {
        methods: ["GET"],
        path: "/buildings/{buildID}/floors",
        produces: ["application/json"]
    }
    documentation {
        ### Get floors for a specific building ID
        A resource is an invokable API method
        Accessible at '/sims/building/buildID/floor'
    }
    getFloorsByBuilding (endpoint caller, http:Request request, string buildID) {

        http:Response response = new;

        // Validate input is an integer
        if(isInteger(buildID)){
            sql:Parameter _buildID = {sqlType:sql:TYPE_VARCHAR, value:sanitizeInt(buildID)};
        
            var selectRet = simsDB->select("select * from tbl_SIMS_Floor where building = (select [name] from tbl_SIMS_Building where Building_Number = ? )", Floor, _buildID);
            table<Floor> dt;

            match selectRet {
            table tableReturned => dt = tableReturned;
            error e => log:printError("Select data from tbl_SIMS_Floor failed: "
                                + e.message);
            }

            

            var jsonConversionRet = <json>dt;
            match jsonConversionRet {
                json jsonRes => {
                    log:printTrace(io:sprintf("%s", jsonRes));
                    
                    // Untainting return as this resource takes no user input
                    response.setJsonPayload(untaint jsonRes, contentType = "application/json");
                    _ = caller -> respond(response);

                }
                error e => {
                    json errResponse = {
                        "status" : "Connection Error",
                        "error": "500 - Internal Server Error"
                    };
                    response.statusCode = 500;
                    response.setJsonPayload(untaint errResponse, contentType = "application/json");
                    _ = caller -> respond(response);
                    log:printError(e.message);
                }
            }
        }
        else {
            json json400 = {
                "status": "Invalid building ID",
                "error": "400 - Bad Request"
            };
            response.statusCode = 400;
            response.setJsonPayload(json400, contentType = "application/json");
            _ = caller -> respond(response);
            log:printInfo(json400.toString());
        }
    }
    @http:ResourceConfig {
        methods: ["GET"],
        path: "/buildings/{buildID}/rooms",
        produces: ["application/json"]
    }
    documentation {
        ### Get rooms for a specific building ID
        A resource is an invokable API method
        Accessible at '/sims/building/buildID/rooms'
    }
    getRoomsByBuilding (endpoint caller, http:Request request, string buildID) {

        http:Response response = new;

        // Validate input is an integer
        if(isInteger(buildID)){
            sql:Parameter _buildID = {sqlType:sql:TYPE_VARCHAR, value:sanitizeInt(buildID)};
        
            var selectRet = simsDB->select("select * from tbl_SIMS_Room where building = (select [name] from tbl_SIMS_Building where Building_Number = ? )", Room, _buildID);
            table<Room> dt;

            match selectRet {
            table tableReturned => dt = tableReturned;
            error e => log:printError("Select data from tbl_SIMS_Room failed: "
                                + e.message);
            }

            

            var jsonConversionRet = <json>dt;
            match jsonConversionRet {
                json jsonRes => {
                    log:printTrace(io:sprintf("%s", jsonRes));
                    
                    // Untainting return as this resource takes no user input
                    response.setJsonPayload(untaint jsonRes, contentType = "application/json");
                    _ = caller -> respond(response);

                }
                error e => {
                    json errResponse = {
                        "status" : "Connection Error",
                        "error": "500 - Internal Server Error"
                    };
                    response.statusCode = 500;
                    response.setJsonPayload(untaint errResponse, contentType = "application/json");
                    _ = caller -> respond(response);
                    log:printError(e.message);
                }
            }
        }
        else {
            json json400 = {
                "status": "Invalid building ID",
                "error": "400 - Bad Request"
            };
            response.statusCode = 400;
            response.setJsonPayload(json400, contentType = "application/json");
            _ = caller -> respond(response);
            log:printInfo(json400.toString());
            log:printInfo(request.rawPath);
        }
    }
    @http:ResourceConfig {
        methods: ["GET"],
        path: "/rooms",
        produces: ["application/json"]
    }
    documentation {
        ### Get rooms for a specific building ID
        A resource is an invokable API method
        Accessible at '/sims/rooms'
    }
    getAllRooms (endpoint caller, http:Request request) {

        http:Response response = new;

        // Validate input is an integer
        
        // /sims/rooms/_search?filter_path=hits.hits.*
        // tune es instance max query size to 200k
        var elasticResp = elastic->get("/sims/rooms/_search?filter_path=hits.hits.*&size=120000");

        match elasticResp {
            http:Response  resp => {
                var msg = resp.getJsonPayload();
                io:println("Status Code: ",resp.statusCode);
                io:println(resp);
                match msg {
                    json jsonPayload => {
                        io:println(jsonPayload);
                        // untaint jsonPayload - from trusted ES
                        response.setPayload(untaint jsonPayload);
                        // response.setJsonPayload(untaint jsonPayload, contentType = "application/json");
                        _ = caller -> respond(response);
                    }
                    error err => {
                        // Send safe response to consumer
                        json errResponse = {
                            "status" : "Connection Error",
                            "error": "500 - Internal Server Error"
                            };
                        response.statusCode = 500;
                        response.setJsonPayload(untaint errResponse, contentType = "application/json");
                        _ = caller -> respond(response);
                        // Log explicit error for observablity
                        log:printError(err.message);
                    }
                }
                
            }
            error err => {
                // Send safe response to consumer
                json errResponse = {
                    "status" : "Connection Error",
                    "error": "500 - Internal Server Error"
                    };
                response.statusCode = 500;
                response.setJsonPayload(untaint errResponse, contentType = "application/json");
                _ = caller -> respond(response);
                // Log explicit error for observablity
                log:printError(err.message);
            }
        }
        
    }
    @http:ResourceConfig {
        methods: ["GET"],
        path: "/assets",
        produces: ["application/json"]
    }
    documentation {
        ### Get rooms for a specific building ID
        A resource is an invokable API method
        Accessible at '/sims/assets'
    }
    getAllAssets (endpoint caller, http:Request request) {

        http:Response response = new;

        // Validate input is an integer
        
        // localhost:9200/sims/rooms/_search?filter_path=hits.hits.*
        // tune es instance max query size to 200k
        var selectRet = simsDB->select("select * from tbl_SIMS_Asset", Asset);
        table<Asset> dt;

        match selectRet {
        table tableReturned => dt = tableReturned;
        error e => log:printError("Select data from tbl_SIMS_Room failed: "
                            + e.message);
        }

        

        var jsonConversionRet = <json>dt;
        match jsonConversionRet {
            json jsonRes => {
                log:printTrace(io:sprintf("%s", jsonRes));
                
                // Untainting return as this resource takes no user input
                response.setJsonPayload(untaint jsonRes, contentType = "application/json");
                _ = caller -> respond(response);

            }
            error e => {
                json errResponse = {
                    "status" : "Connection Error",
                    "error": "500 - Internal Server Error"
                };
                response.statusCode = 500;
                response.setJsonPayload(untaint errResponse, contentType = "application/json");
                _ = caller -> respond(response);
                log:printError(e.message);
            }
        }
        
    }
    
}

function sanitizeInt(string input) returns @untainted string {
    string regEx = "[^\\d+$]";
    return input.replace(regEx, "");
}

function isInteger(string input) returns boolean {
    string regEx = "\\d+";
    boolean isInt = check input.matches(regEx);
    return isInt;
}
```

**Steps to reproduce:**

**Affected Versions:**
`Ballerina 0.980.1`

**OS, DB, other environment details and versions:**
`Darwin`

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by SupunS** [comment]
Thanks for reporting the issue. I noticed that there is a dataTable to JSON conversion inside the resource. There was a known issue https://github.com/ballerina-platform/ballerina-lang/issues/9619 in v0.980.1, which causes datatable to JSON conversion to go OOM, if the size of data coming from the DB is very large.

This issue is fixed in the master branch. Is there any possibility for you to try the same scenario with a latest nightly-build (v0.980.2-SNAPSHOT) of ballerina distribution? 

Note that, any JSON read from a datatable works in a streaming manner. So simply converting the dt to JSON and directly sending it back as the response, will not load the anything to memory. However, if we do any modification/printing json (in your case `log:printTrace(io:sprintf("%s", jsonRes));`), then the entire data will be loaded in to memory. So you need to remove that line of code for the streaming to work properly.

**Comment by jimk-osu** [comment]
@SupunS I just locally tested with no code changes using the  `Ballerina 0.980.2-SNAPSHOT` nightly and it still threw `java.lang.OutOfMemoryError: Java heap space`.

The resource that I'm calling on this one is using an http endpoint. In the code its `getAllRooms`.

Should I switch gears on this and try using `io:ByteChannel` as in the [HTTP Streaming](https://ballerina.io/learn/by-example/http-streaming.html) by-example?

**Comment by jimk-osu** [comment]
I'm currently trying to use `stream<record>` but hitting some hurdles mapping to the refined elasticsearch response:

```json
{
    "hits": {
        "hits": [
            {
                "_index": "sims",
                "_type": "rooms",
                "_id": "sUkE3WQBgbOD3_8hqY_X",
                "_score": 1.0,
                "_source": {
                    "building": "building",
                    "function": "som func",
                    "earning_operation": "false",
                    "capacity": "4.0",
                    "POM_utility": "false",
                    "floor": "02",
                    "recordID": "1234",
                    "space_ID": "1234",
                    "area": "373.33",
                    "room_number": "0651",
                    "comment": "None",
                    "room_Type": "9C-Apartment",
                    "effective_date": "None",
                    "assignment_number": "None",
                    "POM_maintenance": "false",
                    "department": "Some Department",
                    "organization": "Some Org",
                    "sub_room_type": "Bedroom within Suite",
                    "POM_custodial": "false",
                    "type": "None",
                    "campus": "This Campus"
                }
            }
        ]
    }
}
```

**Comment by SupunS** [comment]
When we say `var msg = resp.getJsonPayload();` it will load the entire payload in to memory. You can simply respond back to the client with the same payload without loading it to the memory like below:
```ballerina
var elasticResp = elastic->get("/sims/rooms/_search?filter_path=hits.hits.*&size=120000");

match elasticResp {
    http:Response  resp => {
        _ = caller -> respond(resp);
    }
    error err => {
            // handle error
    }
}
```

Here we simply passthrough the response from the endpoint back to the client, without really touching it.


On a side note, seems like loading the JSON to memory eats up too much memory than it suppose to. We are further investigating this.

**Comment by jimk-osu** [comment]
@SupunS The passthrough works great. From a consumer perspective we'll still need to do something to filter out all the elasticsearch metadata so excited to hear about the outcome of handling large JSON in memory.

In regards to the SQL streaming comment you made, using the nightly `Ballerina 0.980.2-SNAPSHOT` and commenting out the trace statement the response to the call is :
```
[]
```

**Comment by Manuri** [comment]
@jimk-osu We are checking on the empty payload receiving issue.

**Comment by SupunS** [comment]
@jimk-osu We have fixed the issue #9870. This will be available in the next release. Thanks for catching that!

With respect to the JSON object model, we have created a separate issue https://github.com/ballerina-platform/ballerina-lang/issues/9849 to track this. As of now, we are considering that as a possible improvement to the existing model. So even with that improvement, it will consume a considerable amount of memory if you try to load the entire JSON to memory and do filtering on top of it.
